File: python/day_10.py
# -*- coding: utf-8 -*-

from collections import namedtuple

data = None
with open('../resources/day10.in', 'r') as file:
    data = file.read().split("\n")

matrix = [ [int(charr) for charr in line] for line in data if len(line) > 0]

MAX_X = len(matrix[0])
MAX_Y = len(matrix)


Point = namedtuple('Point', ['x', 'y', 'value'])


def map_starting_points():
    starting_points = []

    for i in range(len(matrix)):
        for j in range(len(matrix[i])):
            if matrix[i][j] == 0:
                starting_points.append(Point(j, i, 0))

    return starting_points


def in_bound(x, y):
    return x >= 0 and x < MAX_X and y >= 0 and y < MAX_Y


def get_safe_coord(x, y):
    return (x, y) if in_bound(x, y) else None


def next_coord(x, y, direction):
    if 'up' == direction:
        return get_safe_coord(x, y - 1)

    elif 'down' == direction:
        return get_safe_coord(x, y + 1)

    elif 'right' == direction:
        return get_safe_coord(x + 1, y)

    elif 'left' == direction:
        return get_safe_coord(x - 1, y)

    raise Exception(f"Unknown direction {direction}")


def walk(start):
    x, y = (start.x, start.y)
    possible_points = []

    for direction in ['up', 'right', 'down', 'left']:
        coord = next_coord(x, y, direction)

        if not coord:
            continue

        c_x, c_y = coord
        value = matrix[c_y][c_x]

        point = Point(c_x, c_y, value)

        if (point.value - start.value) == 1:
            possible_points.append(point)

    return possible_points


def part_one():
    starts = map_starting_points()
    score = 0

    for start in starts:
        processing_queue = []
        point = start
        visited = set()

        while True:
            processing_queue = walk(point) + processing_queue

            if len(processing_queue) == 0:
                break

            point = processing_queue.pop(0)

            if (start, point) in visited:
                continue

            if point.value == 9:
                score += 1

            visited.add((start, point))

    return score


print(f"Part 1: {part_one()}")

File: js/day_10.js
'use strict';

import fs from 'fs';

const data = fs.readFileSync('../resources/day10.in', 'utf-8');

const matrix = data.split("\n").map((line) => {
  return line.split('').map((e) => parseInt(e));
}).filter((e) => e.length > 0);


const MAX_X = matrix[0].length;
const MAX_Y = matrix.length;


function mapTrailHeads() {
  const result = [];
  for (let y = 0; y < matrix.length; y++) {
    for (let x = 0; x < matrix[y].length; x++) {
      const value = matrix[y][x];

      if (value === 0)
        result.push([x, y, 0]);
    }
  }

  return result;
}


function isInBound(x, y) {
  return x >= 0 && x < MAX_X && y >= 0 && y < MAX_Y;
}


function getSafeCoord(x, y) {
  return isInBound(x, y) ? [x, y, matrix[y][x]] : null;
}


function nextCoord(x, y, direction) {
  if ('up' === direction)
    return getSafeCoord(x, y - 1);

  else if ('down' === direction)
    return getSafeCoord(x, y + 1);

  else if ('right' === direction)
    return getSafeCoord(x + 1, y);

  else if ('left' === direction)
    return getSafeCoord(x - 1, y);

  throw new Error(`Unknown direction ${direction}`);
}


function walk(start) {
  const [x, y, value] = start;
  const possiblePoints = [];
  const directions = ['up', 'right', 'down', 'left'];

  for (let i = 0; i < directions.length; i++) {
    const coord = nextCoord(x, y, directions[i]);

    if (coord == null) {
      continue;
    }

    if ((coord[2] - value) == 1) {
      possiblePoints.push(coord);
    }

  }

  return possiblePoints;
}


function walkTrailhead(start, unique) {
  let score = 0;
  let processingQueue = [];
  let point = start;
  const visited = new Set();

  while(true) {
    processingQueue = walk(point).concat(processingQueue);

    if (processingQueue.length === 0)
      break;

    point = processingQueue.shift();

    if (unique && visited.has(`${point}`))
      continue;

    if (point[2] === 9)
      score++;

    visited.add(`${point}`);
  }

  return score;
}


function walkAllTrailheads(unique) {
  return mapTrailHeads()
    .map((start) => walkTrailhead(start, unique))
    .reduce((acc, number) => acc + number);
}


function partOne() {
  return walkAllTrailheads(true);
}


function partTwo() {
  return walkAllTrailheads(false);
}

console.log(`Part 1: ${partOne()}`);
console.log(`Part 2: ${partTwo()}`);
